### PR TITLE
GetUserData functions

### PR DESCRIPTION
**Pull request by gardner48**
* Add `GetUserData` functions to each package 
* See proposal in #85

### CODE REVIEW

File: doc/arkode/guide/source/Usage/ARKStep_c_interface/User_callable.rst [hunk 1]
```diff
@@ -3596,6 +3597,21 @@ Number of constraint test failures                     :c:func:`ARKStepGetNumCon
...
 
 
+.. c:function:: int ARKStepGetUserData(void* arkode_mem, void** user_data)
+
+   Returns the user data pointer previously set with
+   :c:func:`ARKStepSetUserData`.
+
+   **Arguments:**
+      * *arkode_mem* -- pointer to the ARKStep memory block.
+      * *user_data* -- pointer to a user data pointer.
```
Comment (drreynolds) [inline doc/arkode/guide/source/Usage/ARKStep_c_interface/User_callable.rst:3607]: Would this read more clearly as "memory reference to a user data pointer"?

File: doc/arkode/guide/source/Usage/ERKStep_c_interface/User_callable.rst [hunk 2]
```diff
@@ -1960,6 +1961,20 @@ Main solver optional output functions
...
 
 
+.. c:function:: int ERKStepGetUserData(void* arkode_mem, void** user_data)
+
+   Returns the user data pointer previously set with
+   :c:func:`ERKStepSetUserData`.
+
+   **Arguments:**
+      * *arkode_mem* -- pointer to the ERKStep memory block.
+      * *user_data* -- pointer to a user data pointer.
```
Comment (drreynolds) [inline doc/arkode/guide/source/Usage/ERKStep_c_interface/User_callable.rst:1971]: Would this read more clearly as "memory reference to a user data pointer"?

File: doc/arkode/guide/source/Usage/MRIStep_c_interface/User_callable.rst [hunk 3]
```diff
@@ -2728,6 +2730,22 @@ Main solver optional output functions
...
+.. c:function:: int MRIStepGetUserData(void* arkode_mem, void** user_data)
+
+   Returns the user data pointer previously set with
+   :c:func:`MRIStepSetUserData`.
+
+   **Arguments:**
+
+   * *arkode_mem* -- pointer to the MRIStep memory block.
+
+   * *user_data* -- pointer to a user data pointer.
```
Comment (drreynolds) [inline doc/arkode/guide/source/Usage/MRIStep_c_interface/User_callable.rst:2742]: Would this read more clearly as "memory reference to a user data pointer"?

File: doc/cvode/guide/source/Usage/index.rst [hunk 4]
```diff
@@ -2660,6 +2662,20 @@ described next.
...
 
 
+.. c:function:: int CVodeGetUserData(void* cvode_mem, void** user_data)
+
+   The function ``CVodeGetUserData`` returns the user data pointer provided to
+   :c:func:`CVodeSetUserData`.
+
+   **Arguments:**
+     * ``cvode_mem`` -- pointer to the CVODE memory block.
+     * ``user_data`` -- pointer to the user data pointer.
```
Comment (drreynolds) [inline doc/cvode/guide/source/Usage/index.rst:2672]: Would this read more clearly as "memory reference to a user data pointer"?

File: doc/cvodes/guide/source/Usage/SIM.rst [hunk 5]
```diff
@@ -2667,6 +2669,20 @@ described next.
...
 
 
+.. c:function:: int CVodeGetUserData(void* cvode_mem, void** user_data)
+
+   The function ``CVodeGetUserData`` returns the user data pointer provided to
+   :c:func:`CVodeSetUserData`.
+
+   **Arguments:**
+     * ``cvode_mem`` -- pointer to the CVODES memory block.
+     * ``user_data`` -- pointer to the user data pointer.
```
Comment (drreynolds) [inline doc/cvodes/guide/source/Usage/SIM.rst:2679]: Would this read more clearly as "memory reference to a user data pointer"?

File: doc/ida/guide/source/Usage/index.rst [hunk 6]
```diff
@@ -2629,6 +2635,19 @@ described next.
...
       * ``IDA_MEM_FAIL`` -- The ``SUNNonlinearSolver`` object is ``NULL``.
 
+.. c:function:: int IDAGetUserData(void* ida_mem, void** user_data)
+
+   The function ``IDAGetUserData`` returns the user data pointer provided to
+   :c:func:`IDASetUserData`.
+
+   **Arguments:**
+     * ``ida_mem`` -- pointer to the IDA memory block.
+     * ``user_data`` -- pointer to the user data pointer.
```
Comment (drreynolds) [inline doc/ida/guide/source/Usage/index.rst:2645]: Would this read more clearly as "memory reference to a user data pointer"?

File: doc/idas/guide/source/Usage/SIM.rst [hunk 7]
```diff
@@ -2663,6 +2669,19 @@ described next.
...
       * ``IDA_MEM_FAIL`` -- The ``SUNNonlinearSolver`` object is ``NULL``.
 
+.. c:function:: int IDAGetUserData(void* ida_mem, void** user_data)
+
+   The function ``IDAGetUserData`` returns the user data pointer provided to
+   :c:func:`IDASetUserData`.
+
+   **Arguments:**
+     * ``ida_mem`` -- pointer to the IDAS memory block.
+     * ``user_data`` -- pointer to the user data pointer.
```
Comment (drreynolds) [inline doc/idas/guide/source/Usage/SIM.rst:2679]: Would this read more clearly as "memory reference to a user data pointer"?

File: doc/kinsol/guide/source/Usage/index.rst [hunk 8]
```diff
@@ -1595,6 +1600,20 @@ functions are described next.
...
 
 
+.. c:function:: int KINGetUserData(void* kin_mem, void** user_data)
+
+   The function :c:func:`KINGetUserData` returns the user data pointer provided
+   to :c:func:`KINSetUserData`.
+
+   **Arguments:**
+     * ``kin_mem`` -- pointer to the KINSOL memory block.
+     * ``user_data`` -- pointer to the user data pointer.
```
Comment (drreynolds) [inline doc/kinsol/guide/source/Usage/index.rst:1610]: Would this read more clearly as "memory reference to a user data pointer"?

File: include/idas/idas.h [hunk 9]
```diff
@@ -317,6 +317,7 @@ SUNDIALS_EXPORT int IDAGetNonlinSolvStats(void *ida_mem, long int *nniters,
                                           long int *nnfails);
 SUNDIALS_EXPORT int IDAGetNumStepSolveFails(void *ida_mem,
                                             long int *nncfails);
+SUNDIALS_EXPORT int IDAGetUserData(void *cvode_mem, void** user_data);
```
Comment (drreynolds) [inline include/idas/idas.h:320]: Should be `void *ida_mem`

File: include/ida/ida.h [hunk 10]
```diff
@@ -232,6 +232,7 @@ SUNDIALS_EXPORT int IDAGetNonlinSolvStats(void *ida_mem, long int *nniters,
                                           long int *nnfails);
 SUNDIALS_EXPORT int IDAGetNumStepSolveFails(void *ida_mem,
                                             long int *nncfails);
+SUNDIALS_EXPORT int IDAGetUserData(void *cvode_mem, void** user_data);
```
Comment (drreynolds) [inline include/ida/ida.h:235]: Should be `void *ida_mem`

File: include/kinsol/kinsol.h [hunk 11]
```diff
@@ -155,6 +155,7 @@ SUNDIALS_EXPORT int KINGetNumBetaCondFails(void *kinmem, long int *nbcfails);
 SUNDIALS_EXPORT int KINGetNumBacktrackOps(void *kinmem, long int *nbacktr);
 SUNDIALS_EXPORT int KINGetFuncNorm(void *kinmem, realtype *fnorm);
 SUNDIALS_EXPORT int KINGetStepLength(void *kinmem, realtype *steplength);
+SUNDIALS_EXPORT int KINGetUserData(void *cvode_mem, void** user_data);
```
Comment (drreynolds) [inline include/kinsol/kinsol.h:158]: Should be `void *kinmem`

File: src/arkode/arkode_io.c [hunk 12]
```diff
@@ -1646,6 +1646,27 @@ int arkGetNumStepSolveFails(void *arkode_mem, long int *nncfails)
...
+/*---------------------------------------------------------------
+  arkGetUserData:
+
+  Returns the user data pointer
+  ---------------------------------------------------------------*/
+int arkGetUserData(void *arkode_mem, void** user_data)
+{
+  ARKodeMem ark_mem;
+  if (arkode_mem==NULL) {
+    arkProcessError(NULL, ARK_MEM_NULL, "ARKode",
```
Comment (drreynolds) [inline src/arkode/arkode_io.c:1658]: Should be "ARKODE"

File: src/idas/idas_io.c [hunk 13]
```diff
@@ -2057,6 +2057,24 @@ int IDAGetNumStepSensSolveFails(void *ida_mem, long int *nSncfails)
 
 /*-----------------------------------------------------------------*/
 
+int IDAGetUserData(void *ida_mem, void** user_data)
+{
+  IDAMem IDA_mem;
+
+  if (ida_mem==NULL) {
+    IDAProcessError(NULL, IDA_MEM_NULL, "IDASs", "IDAGetUserData", MSG_NO_MEM);
```
Comment (drreynolds) [inline src/idas/idas_io.c:2065]: Should be "IDAS"

File: test/unit_tests/arkode/C_serial/ark_test_getuserdata.c [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+  /* Create ARKODE mem structure */
+  arkode_mem = MRIStepCreate(f, NULL, ZERO, y, inner_stepper, sunctx);
+  if (!arkode_mem)
+  {
+    fprintf(stderr, "ARKStepCreate returned NULL\n");
+    return 1;
+  }
+
+  /* Set user data */
+  retval = ARKStepSetUserData(arkode_mem, &udata_in);
```
Comment (drreynolds) [inline test/unit_tests/arkode/C_serial/ark_test_getuserdata.c:185]: This should call `MRIStepSetUserData()`

File: test/unit_tests/arkode/C_serial/ark_test_getuserdata.c [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+  if (retval)
+  {
+    fprintf(stderr, "ARKStepSetUserData returned %i\n", retval);
+    return 1;
+  }
+
+  /* Get user data */
+  retval = ARKStepGetUserData(arkode_mem, &udata_out);
```
Comment (drreynolds) [inline test/unit_tests/arkode/C_serial/ark_test_getuserdata.c:193]: This should call `MRIStepGetUserData()`

**Review by drreynolds — CHANGES_REQUESTED** [review]
A few minor changes are needed, along with some suggested documentation phrasing changes (not required).

**Review by balos1 — APPROVED** [review]

**Pull request merged**